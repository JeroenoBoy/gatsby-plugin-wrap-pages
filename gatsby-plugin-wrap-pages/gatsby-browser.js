export { wrapPageElement } from './wrap-page'
