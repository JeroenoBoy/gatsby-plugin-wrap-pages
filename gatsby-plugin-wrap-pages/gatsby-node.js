const systemPath = require('path')
const fs = require('fs')
const chokidar = require('chokidar')
const { slash } = require('gatsby-core-utils')

const DEFAULT_WRAPPER_NAME = 'wrap-pages'
let ROOT_PATH = null
let pagesToAddContext = {}
let scopeFiles = {}

exports.onCreateWebpackConfig = ({ actions, plugins, store }) => {
  // const root = store.getState().program.directory
  const cacheFilePath = systemPath.resolve(ROOT_PATH, '.cache/wpe-scopes.js')
  actions.setWebpackConfig({
    plugins: [
      plugins.define({
        WDE_CACHE_PATH: JSON.stringify(cacheFilePath),
      }),
    ],
  })
}

exports.onPostBootstrap = async (
  { store, actions, createContentDigest },
  { wrapperName = null } = {}
) => {
  const { program, pages } = store.getState()
  ROOT_PATH = program.directory

  await preparePages({ pages, actions, createContentDigest, wrapperName })
}

exports.onCreateDevServer = (
  { store, actions, emitter, createContentDigest },
  { wrapperName = null } = {}
) => {
  const updatePages = async ({ filterPage = null, filterDir = null }) => {
    const { pages } = store.getState()

    await preparePages({
      pages,
      actions,
      createContentDigest,
      wrapperName,
      filterPage,
      filterDir,
    })
  }

  emitter.on('CREATE_PAGE', async (action) => {
    // const oldCache = require.cache[require.resolve(componentPath)]
    const filterPage = slash(action.payload.component)
    await updatePages({ filterPage })
  })

  const watchPath = systemPath.join(
    ROOT_PATH,
    '**',
    wrapperName || DEFAULT_WRAPPER_NAME + '*'
  )
  const watcher = chokidar.watch(watchPath, () => {})

  watcher.on('unlink', async (path) => {
    await updatePages({ filterDir: systemPath.dirname(path) })
  })

  // delay the watcher, because the "add" event gets emitted right away
  setTimeout(() => {
    watcher.on('watcher add', async (path) => {
      await updatePages({ filterDir: systemPath.dirname(path) })
    })
  }, 100)
}

async function preparePages({
  pages,
  actions,
  createContentDigest,
  wrapperName = null,
  filterPage = null,
  filterDir = null,
}) {
  for (const p of pages) {
    const page = p[1]

    if (filterPage) {
      if (page.component !== filterPage) {
        continue
      }
    }

    if (filterDir) {
      if (systemPath.dirname(page.component) !== filterDir) {
        continue
      }
    }

    if (page.component.includes(wrapperName || '/' + DEFAULT_WRAPPER_NAME)) {
      actions.deletePage(page)
      scopeFiles[systemPath.dirname(page.component)] = page
    } else {
      pagesToAddContext[page.component] = page
    }
  }

  if (filterDir) {
    for (let key in scopeFiles) {
      if (!fs.existsSync(scopeFiles[key].component)) {
        delete scopeFiles[key]
      }
    }
  }

  const wrappersImportPaths = {}

  for (const componentPath in pagesToAddContext) {
    const page = pagesToAddContext[componentPath]
    const dirPath = systemPath.dirname(page.component)

    if (filterPage) {
      if (page.component !== filterPage) {
        continue
      }
    }

    if (filterDir) {
      if (dirPath !== filterDir) {
        continue
      }
    }

    page.context.__WPE__ = page.context.__WPE__ || {}

    const scopePaths = findValidScopePaths(dirPath)

    if (scopePaths.length > 0) {
      for (const scopePath of scopePaths) {
        const isSame = dirPath === scopePath
        const wrapperPage = scopeFiles[scopePath]

        if (wrapperPage) {
          const relativeComponentPath = systemPath.relative(
            ROOT_PATH,
            wrapperPage.component
          )
          const contentDigest = createContentDigest(relativeComponentPath)
          wrappersImportPaths[contentDigest] = relativeComponentPath

          page.context.__WPE__[contentDigest] = { isSame }
        }
      }
    }
  }

  await generateCache({ wrappersImportPaths })
}

async function generateCache({ wrappersImportPaths }) {
  let cacheFileContent = ''
  for (const contentDigest in wrappersImportPaths) {
    const relativeComponentPath = wrappersImportPaths[contentDigest]
    cacheFileContent += `export * as _${contentDigest} from '../${relativeComponentPath}';\n`
  }

  const cacheFilePath = systemPath.resolve(ROOT_PATH, '.cache/wpe-scopes.js')
  await fs.writeFile(cacheFilePath, cacheFileContent)

  /**
When we use "writing to disk", we get this warning:

warn Warning: Event "xstate.after(200)#waitingMachine.ag
gregatingFileChanges" was sent to stopped service
"waitingMachine". This service has already reached its
final state, and will not transition.
Event: {"type":"xstate.after(200)#waitingMachine.aggrega
tingFileChanges"}
*/
}

function findValidScopePaths(componentDir) {
  const result = []
  const parts = componentDir.split(systemPath.sep)

  for (const i of parts) {
    const possibleScope = parts.join(systemPath.sep)

    if (scopeFiles[possibleScope]) {
      result.push(possibleScope)
    }

    // Bail out once we have reached the Gatsby instance root
    if (possibleScope === ROOT_PATH) {
      break
    }

    parts.pop()
  }

  return result
}
