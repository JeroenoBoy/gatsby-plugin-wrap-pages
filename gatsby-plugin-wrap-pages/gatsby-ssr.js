export { wrapPageElement } from './wrap-pages'
