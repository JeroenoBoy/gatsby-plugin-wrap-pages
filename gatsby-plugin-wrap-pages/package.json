{
  "name": "gatsby-plugin-wrap-pages",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "author": "Tobias HÃ¸egh",
  "scripts": {
    "test": "yarn workspace integration-tests test",
    "test:watch": "yarn workspace integration-tests test --watch",
    "lint": "eslint --ext .js --ignore-path ../.gitignore ."
  },
  "peerDependencies": {
    "gatsby": "*",
    "react": "*",
    "react-dom": "*"
  },
  "devDependencies": {
    "@testing-library/react": "^12.0.0",
    "babel-eslint": "^10.1.0",
    "chokidar": "*",
    "eslint": "^7.30.0",
    "eslint-config-react-app": "^6.0.0",
    "eslint-plugin-flowtype": "^5.8.0",
    "eslint-plugin-import": "^2.23.4",
    "eslint-plugin-jest": "^24.3.6",
    "eslint-plugin-jsx-a11y": "^6.4.1",
    "eslint-plugin-react": "^7.24.0",
    "eslint-plugin-react-hooks": "^4.2.0",
    "fs-extra": "*",
    "gatsby-core-utils": "*",
    "jest": "^27.0.6",
    "signal-exit": "*"
  }
}
