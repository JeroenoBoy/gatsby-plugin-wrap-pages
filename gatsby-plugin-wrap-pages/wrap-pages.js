// eslint-disable-next-line
const WCD_SCOPES = require(WDE_CACHE_PATH)

export const wrapPageElement = (params) => {
  try {
    if (WCD_SCOPES) {
      const { WPS } = params.props.pageContext
      if (WPS) {
        for (const { hash, isSame } of WPS) {
          if (hash) {
            const scope = WCD_SCOPES[`_${hash}`]

            if (scope) {
              let name = null

              // 1. wrap current scoped pages
              if (isSame && scope.wrapPages) {
                name = 'wrapPages'
              }

              // 2. wrap current scoped and nested pages
              if (!name) {
                if (scope.wrapPagesDeep) {
                  name = 'wrapPagesDeep'
                } else if (scope.default) {
                  console.info(
                    'gatsby-plugin-wrap-pages: "export default" is not supported yet. If so, shold it wrap deep or only the current scope?'
                  )
                }
              }

              if (scope?.[name]) {
                const result = scope[name](params)
                if (result) {
                  params.element = result
                }
              }
            }
          }
        }
      }
    }
  } catch (e) {
    console.warn('gatsby-plugin-wrap-pages:', e)
  }

  return params.element
}

// NB: can probably be removed as this does not help for the "first save" issue
// delete require.cache[__filename]
