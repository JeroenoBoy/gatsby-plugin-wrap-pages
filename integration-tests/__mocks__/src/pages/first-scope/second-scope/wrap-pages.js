import React from 'react'

export function wrapPages({ element }) {
  return <div data-testid="wrapPages-second">{element}</div>
}

export function wrapPagesDeep({ element }) {
  return <div data-testid="wrapPagesDeep-second">{element}</div>
}
