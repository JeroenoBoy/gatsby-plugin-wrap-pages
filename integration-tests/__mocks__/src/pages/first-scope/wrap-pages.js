import React from 'react'

export function wrapPages({ element }) {
  return <div data-testid="wrapPages-first">{element}</div>
}

export function wrapPagesDeep({ element }) {
  return <div data-testid="wrapPagesDeep-first">{element}</div>
}
