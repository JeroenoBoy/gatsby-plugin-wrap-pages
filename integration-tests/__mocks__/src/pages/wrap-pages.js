import React from 'react'

export function wrapPagesDeep({ element }) {
  return (
    <div data-testid="wrapPagesDeep-main">
      <header role="header">header</header>
      <main role="main">{element}</main>
    </div>
  )
}
