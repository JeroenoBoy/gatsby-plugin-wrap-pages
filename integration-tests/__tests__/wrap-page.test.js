import React from 'react'
import { render, screen } from '@testing-library/react'
import { wrapPageElement } from 'gatsby-plugin-wrap-pages/wrap-pages'

const Page = () => {
  return <div data-testid="content">Page Content</div>
}

describe('wrapPagesDeep', () => {
  beforeEach(() => {
    render(
      wrapPageElement({
        element: <Page />,
        props: {
          pageContext: {
            WPS: [
              { hash: '26243a013ca7d319d87081f9768dd680' },
              { hash: '455a0f696332624bb2ff3f910434400d' },
              { hash: 'de24c938e6d0ae34eea46b0360bc707c' },
            ],
          },
        },
      })
    )
  })

  it('should contain original content', () => {
    const pageElem = screen.getByTestId('content')
    expect(pageElem.textContent).toBe('Page Content')
  })

  it('should wrap additional header', () => {
    const headerElem = screen.getByRole('header')
    expect(headerElem instanceof HTMLElement).toBe(true)
  })

  it('should match snapshot', () => {
    const bodyElem = document.querySelector('body')
    expect(bodyElem).toMatchSnapshot()
  })

  it('should wrap in correct order', () => {
    const elements = document.querySelectorAll('body div[data-testid]')
    expect(elements[0].getAttribute('data-testid')).toBe('wrapPagesDeep-main')
    expect(elements[1].getAttribute('data-testid')).toBe('wrapPagesDeep-first')
    expect(elements[2].getAttribute('data-testid')).toBe('wrapPagesDeep-second')
  })
})

describe('wrapPages on second last page', () => {
  beforeEach(() => {
    render(
      wrapPageElement({
        element: <Page />,
        props: {
          pageContext: {
            WPS: [
              { hash: '26243a013ca7d319d87081f9768dd680' },
              { hash: '455a0f696332624bb2ff3f910434400d', isSame: true },
              { hash: 'de24c938e6d0ae34eea46b0360bc707c' },
            ],
          },
        },
      })
    )
  })

  it('should contain original content', () => {
    const pageElem = screen.getByTestId('content')
    expect(pageElem.textContent).toBe('Page Content')
  })

  it('should wrap additional header', () => {
    const headerElem = screen.getByRole('header')
    expect(headerElem instanceof HTMLElement).toBe(true)
  })

  it('should match snapshot', () => {
    const bodyElem = document.querySelector('body')
    expect(bodyElem).toMatchSnapshot()
  })

  it('should wrap in correct order', () => {
    const elements = document.querySelectorAll('body div[data-testid]')
    expect(elements[0].getAttribute('data-testid')).toBe('wrapPagesDeep-main')
    expect(elements[1].getAttribute('data-testid')).toBe('wrapPages-first')
    expect(elements[2].getAttribute('data-testid')).toBe('wrapPagesDeep-second')
  })
})

describe('wrapPages on last page', () => {
  beforeEach(() => {
    render(
      wrapPageElement({
        element: <Page />,
        props: {
          pageContext: {
            WPS: [
              { hash: '26243a013ca7d319d87081f9768dd680', isSame: true },
              { hash: '455a0f696332624bb2ff3f910434400d' },
              { hash: 'de24c938e6d0ae34eea46b0360bc707c' },
            ],
          },
        },
      })
    )
  })

  it('should wrap in correct order', () => {
    const elements = document.querySelectorAll('body div[data-testid]')
    expect(elements[0].getAttribute('data-testid')).toBe('wrapPagesDeep-main')
    expect(elements[1].getAttribute('data-testid')).toBe('wrapPagesDeep-first')
    expect(elements[2].getAttribute('data-testid')).toBe('wrapPages-second')
  })
})
