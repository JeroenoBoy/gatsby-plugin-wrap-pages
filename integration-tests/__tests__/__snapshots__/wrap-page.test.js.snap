// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapPages on second last page should match snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="wrapPagesDeep-main"
    >
      <header
        role="header"
      >
        header
      </header>
      <main
        role="main"
      >
        <div
          data-testid="wrapPages-first"
        >
          <div
            data-testid="wrapPagesDeep-second"
          >
            <div
              data-testid="content"
            >
              Page Content
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</body>
`;

exports[`wrapPagesDeep should match snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="wrapPagesDeep-main"
    >
      <header
        role="header"
      >
        header
      </header>
      <main
        role="main"
      >
        <div
          data-testid="wrapPagesDeep-first"
        >
          <div
            data-testid="wrapPagesDeep-second"
          >
            <div
              data-testid="content"
            >
              Page Content
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</body>
`;
